Lib/pdb.py

```diff
@@ -545,15 +547,71 @@ def _disable_tab_completion(self):
         else:
             yield
 
+    def _exec_in_closure(self, source, globals, locals):
+        """ Run source code in closure so code object created within source
+            can find variables in locals correctly
+        """
+
+        # If the source is an expression, we need to print its value
+        try:
+            compile(source, "<string>", "eval")
+            source = "__pdb_eval_result__ = " + source
+            locals["__pdb_eval_result__"] = None
+        except SyntaxError:
+            pass
+
+        # Add write-back to update the locals
+        locals["__pdb_write_back__"] = {}
+        source = ("try:\n" +
+                  textwrap.indent(source, "  ") + "\n" +
+                  "finally:\n" +
+                  "  for key, val in locals().items():\n    __pdb_write_back__[key] = val")
+
+        try:
+            local_vars = list(locals.keys())
+
+            # Build a closure source code with freevars from locals like:
+            # def outer():
+            #   var = None
+            #   def __pdb_scope():  # This is the code object we want to execute
+            #     nonlocal var
+            #     <source>
+            source_with_closure = ("def outer():\n" +
+                                   "\n".join(f"  {var} = None" for var in local_vars) + "\n" +
+                                   "  def __pdb_scope():\n" +
+                                   "\n".join(f"    nonlocal {var}" for var in local_vars) + "\n" +
+                                   textwrap.indent(source, "    ")
+                                   )
+
+            # Compile the instrumented source code, and get the code object of __pdb_scope()
+            # This simulates executing the original source code with locals as cellvars
+            # co_consts[0] -> outer;
+            # outer.co_consts[1] -> __pdb_scope because outer.co_consts[0] is None
+            code = compile(source_with_closure, "<string>", "exec").co_consts[0].co_consts[1]
```

Comment:
I think we can do this without rooting around in `co_consts`:

```suggestion
            # Build a closure source code with freevars from locals like:
            # def __pdb_nonlocal_scope__():
            #   var = None
            #   def __pdb_local_scope__():  # This is the code object we want to execute
            #     nonlocal var
            #     <source>
            #   return __pdb_local_scope__.__code__
            source_with_closure = ("def __pdb_nonlocal_scope__():\n" +
                                   "\n".join(f"  {var} = None" for var in local_vars) + "\n" +
                                   "  def __pdb_local_scope__():\n" +
                                   "\n".join(f"    nonlocal {var}" for var in local_vars) + "\n" +
                                   textwrap.indent(source, "    ") +
                                   "\n  return __pdb_local_scope__.__code__"
                                   )

            # Compile the instrumented source code, and get the code object of __pdb_scope()
            # This simulates executing the original source code with locals as cellvars:
            code = exec(source_with_closure, globals, locals)
            code = locals.pop("__pdb_nonlocal_scope__")()
```

Reply:
Is this correct? The code you got is a function I think? And you are supposed to get the return value of the function?

Reply:
I modified the outer function to return the inner function’s code. Then we fish the outer function out of the globals, and call it to get the code.

Reply:
Where did you call the outer function?

Reply:
`code = locals.pop("__pdb_nonlocal_scope__")()`. Note the trailing parens.

I guess you could also call it in the source, then save the result in a local. This just seemed cleaner.